hypothesis-python/src/hypothesis/extra/_array_helpers.py

```diff
@@ -0,0 +1,681 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2021 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+import re
+from typing import NamedTuple, Optional, Tuple, Union
+
+from hypothesis import assume, strategies as st
+from hypothesis.errors import InvalidArgument
+from hypothesis.internal.conjecture import utils as cu
+from hypothesis.internal.coverage import check_function
+from hypothesis.internal.validation import check_type, check_valid_interval
+from hypothesis.strategies._internal.utils import defines_strategy
+from hypothesis.utils.conventions import UniqueIdentifier, not_set
+
+__all__ = [
+    "NDIM_MAX",
+    "Shape",
+    "BroadcastableShapes",
+    "BasicIndex",
+    "check_argument",
+    "order_check",
+    "check_valid_dims",
+    "array_shapes",
+    "valid_tuple_axes",
+    "broadcastable_shapes",
+    "mutually_broadcastable_shapes",
+    "MutuallyBroadcastableShapesStrategy",
+    "BasicIndexStrategy",
+]
+
+
+Shape = Tuple[int, ...]
+# We silence flake8 here because it disagrees with mypy about `ellipsis` (`type(...)`)
+BasicIndex = Tuple[Union[int, slice, None, "ellipsis"], ...]  # noqa: F821
+
+
+class BroadcastableShapes(NamedTuple):
+    input_shapes: Tuple[Shape, ...]
+    result_shape: Shape
+
+
+@check_function
+def check_argument(condition, fail_message, *f_args, **f_kwargs):
+    if not condition:
+        raise InvalidArgument(fail_message.format(*f_args, **f_kwargs))
+
+
+@check_function
+def order_check(name, floor, min_, max_):
+    if floor > min_:
+        raise InvalidArgument(f"min_{name} must be at least {floor} but was {min_}")
+    if min_ > max_:
+        raise InvalidArgument(f"min_{name}={min_} is larger than max_{name}={max_}")
+
+
+# 32 is a dimension limit specific to NumPy, and does not necessarily apply to
+# other array/tensor libraries. Historically these strategies were built for the
+# NumPy extra, so it's nice to keep these limits, and it's seemingly unlikely
+# someone would want to generate >32 dim arrays anyway.
+# See https://github.com/HypothesisWorks/hypothesis/pull/3067.
+NDIM_MAX = 32
+
+
+@check_function
+def check_valid_dims(dims, name):
+    if dims > NDIM_MAX:
+        raise InvalidArgument(
+            f"{name}={dims}, but Hypothesis does not support arrays with "
+            f"dimensions greater than {NDIM_MAX}"
```

Comment:
```suggestion
            f"more than {NDIM_MAX} dimensions"
```

Could otherwise be confused with a limitation on side length.